src/components/fields/ArrayField.js

```diff
@@ -185,6 +205,51 @@ class ArrayField extends Component {
     autofocus: false,
   };
 
+  constructor(props) {
+    super(props);
+    const { formData } = props;
+    const keyedFormData = generateKeyedFormData(formData);
+    this.state = {
+      keyedFormData,
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const nextFormData = nextProps.formData;
+    const previousKeyedFormData = this.state.keyedFormData;
+    const newKeyedFormData =
+      nextFormData.length === previousKeyedFormData.length
+        ? previousKeyedFormData.map((previousKeyedFormDatum, index) => {
+            return {
+              key: previousKeyedFormDatum.key,
+              item: nextFormData[index],
+            };
+          })
+        : generateKeyedFormData(nextFormData);
```

Comment:
Won't this reset _all_ the keys when an array element is added? Do you think that's ok behavior?

Reply:
If insertion is done via onAddClick, the newly added entry will already be in the state with an assigned key. If insertion is done externally, then it is likely safer to reassign keys as there's no way to reconcile any other possible changes in the form data with that in the state.

Reply:
Makes sense, thanks!